bundles/org.openhab.binding.icalendar/src/main/java/org/openhab/binding/icalendar/internal/logic/BiweeklyPresentableCalendar.java

```diff
@@ -0,0 +1,322 @@
+                break;
+            }
+        }
+        return eventList;
+    }
+
+    @Override
+    public @Nullable List<Event> getJustEndedEvents(Instant frameBegin, Instant frameEnd) {
+        List<Event> eventList = null;
+        // process all the events in the iCalendar
+        for (VEvent event : usedCalendar.getEvents()) {
+            Duration duration = getEventLength(event);
+            if (duration == null) {
+                continue;
+            }
+            // iterate over all begin dates
+            DateIterator begDates = getRecurredEventDateIterator(event);
+            while (begDates.hasNext()) {
+                Instant begInst = begDates.next().toInstant();
+                Instant endInst = begInst.plus(duration);
+                if (endInst.isBefore(frameBegin)) {
+                    continue;
+                } else if (endInst.isAfter(frameEnd)) {
+                    break;
+                }
+                // fall through => means we are within the time frame
+                if (eventList == null) {
+                    eventList = new ArrayList<Event>();
+                }
+                eventList.add(new VEventWPeriod(event, begInst, endInst).toEvent());
+                break;
+            }
+        }
+        return eventList;
+    }
+
+    @Override
+    public @Nullable Event getNextEvent(Instant instant) {
+        Collection<VEventWPeriod> candidates = new ArrayList<VEventWPeriod>();
+        Collection<VEvent> negativeEvents = new ArrayList<VEvent>();
+        Collection<VEvent> positiveEvents = new ArrayList<VEvent>();
+        classifyEvents(positiveEvents, negativeEvents);
+        for (VEvent currentEvent : positiveEvents) {
+            DateIterator startDates = this.getRecurredEventDateIterator(currentEvent);
+            Duration duration = getEventLength(currentEvent);
+            if (duration == null) {
+                continue;
+            }
+            startDates.advanceTo(Date.from(instant));
+            while (startDates.hasNext()) {
+                Instant startInstant = startDates.next().toInstant();
+                if (startInstant.isAfter(instant)) {
+                    @Nullable
+                    Uid currentEventUid = currentEvent.getUid();
+                    if (currentEventUid == null || !isCounteredBy(startInstant, currentEventUid, negativeEvents)) {
+                        candidates.add(new VEventWPeriod(currentEvent, startInstant, startInstant.plus(duration)));
+                        break;
+                    }
+                }
+            }
+        }
+        VEventWPeriod earliestNextEvent = null;
+        for (VEventWPeriod positiveCandidate : candidates) {
+            if (earliestNextEvent == null || earliestNextEvent.start.isAfter(positiveCandidate.start)) {
+                earliestNextEvent = positiveCandidate;
+            }
+        }
+
```

Comment:
This is not possible as `candidates` is of type `Collection<@NonNull VEventPeriod>`. The API is new to me - is there a workaround?

Reply:
I'm not sure what you mean. I can't do a `orElse(null)` because `candidates`' Collection-type (and it's resulting stream) has a `@NonNull` implied. the `candidates` Collection can also be empty.